vault/resource_identity_entity_alias.go

```diff
@@ -63,9 +63,22 @@ func identityEntityAliasCreate(d *schema.ResourceData, meta interface{}) error {
 	if err != nil {
 		return fmt.Errorf("error writing IdentityEntityAlias to %q: %s", name, err)
 	}
-	log.Printf("[DEBUG] Wrote IdentityEntityAlias %q", name)
 
-	d.SetId(resp.Data["id"].(string))
+	if resp != nil {
+		d.SetId(resp.Data["id"].(string))
+	} else {
+		// A nil response is possible if the alias already exists and TF doesn't know about it.
+		// But we still need and don't have the alias ID, so a search is used to find it.
+		log.Printf("[DEBUG] Entity alias ID is unknown. Reading aliases for entity %q", canonicalID)
+
+		id, err := findAliasID(client, canonicalID, name, mountAccessor)
+		if err != nil {
+			return fmt.Errorf("error writing IdentityEntityAlias to %q: %s", name, err)
+		}
+		d.SetId(id)
+	}
```

Comment:
This part is a bit magical from the end user perspective. From a Vault perspective this sounds great, it's fault tolerant. From a Terraform perspective, the user _thinks_ they just created the thing, but it actually already exists. 

Assuming the response is `nil` _because_ the alias already exists, would it be possible to check the response status code and make a decision? If it's `204` (I assume this means the alias exists) maybe we should error here, and tell the user this thing already exists. Terraform does have an `import` command, it may be more Terraform-like to error and have them import, than to automagically assume ownership of this thing. What do you think?

Reply:
That is a fair critique and it probably comes down to UX and Terraform norms, which you're more familiar with. We can definitely provide an error message instead if that would be more inline with typically behavior. I've not looked into the `import` support for this resource, though I suspect it would need to be beefed up with much of the same code code (i.e. finding an alias by traits).

Reply:
Could we add the `findAliasID` call to the `READ` method?

```go
func identityEntityAliasRead(d *schema.ResourceData, meta interface{}) error {
  client := meta.(*api.Client)
  id := d.Id()
  if id == "" {
    id,err := findAliasID()
    // err handle
    if id != nil {
      d.SetId(id)
    }
  }
  // other stuff
}
```

That would cover the import functionality, so we could just use `schema.ImportStatePassthrough`:

- https://www.terraform.io/docs/extend/resources/import.html#importer-state-function

The only real difference between these approaches is that one is in `CREATE` and the other is `READ`, and the consequences of "magic" behind the scenes are different to end users. 

If you add configuration for an alias that happens to exist already, then:

`findAliasID` in create:

- Users run `plan`, Terraform refreshes state and plans to create the alias (it wont)
- Users run `apply`, Terraform claims it created the alias (it didn't)


`findAliasID` in read:

- Users run `plan`, Terraform refreshes state and `plan` says there's nothing to do (it added the ID locally to state though)
- Users run `apply`, Terraform doesn't do anything with this resource

Reply:
I'm not sure how the magic in the `import` case would actually work. Import wants an ID to fetch, but we don't have an ID. It complains if you don't provide one, and other ways to fake that check and allow the findAliasID to run get hacky.

I'm thinking we can keep this really simple and if you want to import the alias, provide the ID like normal. You can find the alias ID via the CLI or UI. Then just:

```
terraform import vault_identity_entity_alias.test_alias <alias_id>
```

That seems to work fine, has no magic, and would mean the only change would be to detect that the ID already exists (the `nil` response on Create) and return an error message there.

Reply:
I just pushed an update that retains the alias ID search (which would be a tedious step for the operator trying to locate the ID in order to import or otherwise deal with it), but it only uses the result to provide a more helpful error message.  So, retain the utility, but remove the magic.